src/dmd/func.d

```diff
@@ -1306,7 +1306,14 @@ extern (C++) class FuncDeclaration : Declaration
         if (!fbody)
             return false;
 
-        if (isVirtualMethod())
+        if (isVirtualMethod() &&
+            /*
+             * https://issues.dlang.org/show_bug.cgi?id=21719
+             *
+             * If we have an auto virtual function we can infer
+             * the attributes.
+             */
+            !(inferRetType && !isCtorDeclaration()))
```

Comment:
Maybe add `!isVirtualMethod` below instead of duplicating the logic here?

Reply:
Sorry, but I don't see how that is possible. The following check is for functions that are declared in a function scope and the next check is for literals, instantiated and auto functions. How exactly should `!isVirtualMethod` be added? The only way I can think is to move the check after the check for `auto functions`, but then it is possible that virtual functions inside nested classes will be inferred.

Reply:
Actually, that is even impossible when removing the entire check.
Test case that logs the result of `canInferAttributes`:

```d
class C
{
    final void final_() {}
    void normal() {}
    void safeNormal() @safe {}
    auto virtual() {}
    auto safeVirtual() @safe {}
    auto systemVirtual() @system {}

    class NC
    {
        final void final_() {}
        void normal() {}
        void safeNormal() @safe {}
        auto virtual() {}
        auto safeVirtual() @safe {}
        auto systemVirtual() @system {}
    }
}

extern(C) void main()
{
    class NC
    {
        final void final_() {}
        void normal() {}
        void safeNormal() @safe {}
        auto virtual() {}
        auto safeVirtual() @safe {}
        auto systemVirtual() @system {}
    }
}
```

```
inf.C.final_.canInferAttributes = false
inf.C.normal.canInferAttributes = false
inf.C.safeNormal.canInferAttributes = false
inf.C.virtual.canInferAttributes = true
inf.C.safeVirtual.canInferAttributes = true
inf.C.systemVirtual.canInferAttributes = true

inf.C.NC.final_.canInferAttributes = false
inf.C.NC.normal.canInferAttributes = false
inf.C.NC.safeNormal.canInferAttributes = false
inf.C.NC.virtual.canInferAttributes = true
inf.C.NC.safeVirtual.canInferAttributes = true
inf.C.NC.systemVirtual.canInferAttributes = true

inf.main.canInferAttributes = false

inf.main.NC.final_.canInferAttributes = false
inf.main.NC.normal.canInferAttributes = false
inf.main.NC.safeNormal.canInferAttributes = false
inf.main.NC.virtual.canInferAttributes = true
inf.main.NC.safeVirtual.canInferAttributes = true
inf.main.NC.systemVirtual.canInferAttributes = true
```

Reply:
Why the ctor check though ?

Reply:
I can only assume that constructors do not have an explicit return type and may be considered auto even though they are not. I simply took the check from line [1326](https://github.com/dlang/dmd/pull/13308/files#diff-e88d6d54d409ea665c7643a4a89e3fe01d8decf90042dbf8f11e905a48aabfecR1326)